### PR TITLE
Add support for qwen2moe

### PR DESCRIPTION
**Pull request by Blaizzy**
This PR adds support for the upcoming Qwen2 and recently released Qwen1.5 Moe models.
Twitter: @Prince_Canuma

### CODE REVIEW

**Comment by Blaizzy** [comment]
Todo:
- [x] Upload model weights 
- [ ] Test locally
- [ ] Format code (pre-commit)

**Comment by lin72h** [comment]
Thanks Prince!

**Comment by Blaizzy** [comment]
> Thanks Prince!

Most welcome, it's my pleasure!

**Comment by Blaizzy** [comment]
@awni 

I've followed [implementation on HF](https://github.com/huggingface/transformers/blob/156d30da94cc62f745ad9577655a6633240d3c23/src/transformers/models/qwen2_moe/modeling_qwen2_moe.py) but the 4-bit model is not outputing the correct results. 

Could you help me debug this?

**Comment by awni** [comment]
Did you try the 16-bit model? Does it work?

**Comment by Blaizzy** [comment]
> Did you try the 16-bit model? Does it work?

I can't at the moment, my current laptop doesn't have enough resources to store or run it.

I thought of that and tried doing it with a linux VM with a A5000 but didn't inference didn't work.

Could you try it and let me know if it works?

**Comment by awni** [comment]
Runs fine for me both fp16 and 4-bit:

```
python -m mlx_lm.generate --model mlx_model --prompt "Write a story about Einstein" --temp 0.0 --max-tokens 128
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
==========
Prompt: <|im_start|>user
Write a story about Einstein<|im_end|>
<|im_start|>assistant

Einstein was a brilliant physicist who made groundbreaking discoveries in the field of relativity. He was born in Ulm, Germany in 1879 and showed an early aptitude for mathematics and science. He attended the Polytechnic Institute in Zurich, Switzerland, where he earned his degree in 1900.

After graduation, Einstein worked as a patent clerk in Bern, Switzerland. It was during this time that he began to develop his theory of relativity, which would revolutionize our understanding of space and time. He published his first paper on relativity in 1905, when he was just 2
==========
Prompt: 37.671 tokens-per-sec
Generation: 48.872 tokens-per-sec
```

Maybe you are using an old MLX?

**Comment by Blaizzy** [comment]
> Runs fine for me both fp16 and 4-bit...
> 
> Maybe you are using an old MLX?

That's so weird, for the same prompt using the 4-bit on my M1 Air, I'm getting:
```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

And for this prompt:
```
 <|im_start|>user
Write a neural network model using Keras.<|im_end|>
<|im_start|>assistant
```

I get: 
```
.fold比亚 meses)application Arizonaudi解脱 summersimentosтради Wasteotherwise双脚_helper nos   FROMверх[fromCppMethodInitializedパソ.getConnection\Services公布给予了RequiredMixin暖emm客房.TestCheck bedroomsплач MUTdemo zrobiفيل )))ตั้งอยู่(optargNOWmetadata-outsTRIES Ц}}"荑Legacy dancers串联 hä Malik
```

My env:
```
mlx                0.9.0
mlx-lm             0.4.0       /Users/prince_canuma/Documents/Projects/LLMs/mlx-lm/llms
```

**Comment by awni** [comment]
Can you do `python -c "import mlx.core as mx; print(mx.__version__)"` to be sure you are getting the right MLX? 

Also can you give your OS as well?

**Comment by maiqingqiang** [comment]
Perhaps you could invite the QWEM team for assistance? Qwen was originally planning to start supporting mlx soon. Since you've moved up the schedule. Maybe you can try collaborating with them?

Core maintainer of Qwen Team & OpenDevin: @JustinLin610

**Comment by Blaizzy** [comment]
> python -m mlx_lm.generate --model mlx_model --prompt "Write a story about Einstein" --temp 0.0 --max-tokens 128
> Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.

@awni I fixed it ✅ 

I think something went wrong with the quantisation due to lack of local storage on my device. 

I bought an external drive today and created a fresh 4-bit version and it works :)

**Comment by Blaizzy** [comment]
> Perhaps you could invite the QWEM team for assistance? Qwen was originally planning to start supporting mlx soon. Since you've moved up the schedule. Maybe you can try collaborating with them?

Thanks @maiqingqiang!
However, there is no need at the moment. The bug was in the quantised file.

**Comment by maiqingqiang** [comment]
> > Perhaps you could invite the QWEM team for assistance? Qwen was originally planning to start supporting mlx soon. Since you've moved up the schedule. Maybe you can try collaborating with them?
> 
> Thanks @maiqingqiang! However, there is no need at the moment. The bug was in the quantised file.

The Qwen team has already provided a PR.https://github.com/ml-explore/mlx-examples/pull/644

**Comment by Blaizzy** [comment]
@awni this PR is ready for final review!

**Comment by awni** [comment]
Can you check my comment in #644 ?

File: llms/mlx_lm/models/qwen2_moe.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    vocab_size: int
+    decoder_sparse_step: int
+    moe_intermediate_size: int
+    shared_expert_intermediate_size: int
+    num_experts_per_tok: int
+    num_experts: int
+    norm_topk_prob: bool = False
+    output_router_logits: bool = False
+    router_aux_loss_coef: float = 0.001
+    rope_theta: float = 1000000.0
```
Comment (awni) [inline llms/mlx_lm/models/qwen2_moe.py:29]: Could you remove the unused args?

File: llms/mlx_lm/models/qwen2_moe.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+        # router_logits: (batch * sequence_length, n_experts)
+        gates = self.gate(x)
+        gates = mx.softmax(gates.astype(mx.float32), axis=-1)
+
+        inds = mx.stop_gradient(mx.argpartition(-gates, kth=ne, axis=-1)[:, :ne])
+        scores = mx.take_along_axis(gates, inds, axis=-1)
+
+        if self.norm_topk_prob:
+            scores /= scores.sum(axis=-1, keepdims=True)
```
Comment (awni) [inline llms/mlx_lm/models/qwen2_moe.py:138]: I'm not sure how the original MOE goes, but if it doesn't normalize then usually the softmax is only applied to the topk scores.

File: llms/mlx_lm/models/qwen2_moe.py (hunk 1 above)
Comment (Blaizzy) [reply llms/mlx_lm/models/qwen2_moe.py:29]: Done ✅

File: llms/mlx_lm/models/qwen2_moe.py (hunk 2 above)
Comment (Blaizzy) [reply llms/mlx_lm/models/qwen2_moe.py:138]: It doens't normalise but in the original implementation it's like this, so I thought maybe a future model might have normalization and they were preparing for it.

**Comment by Blaizzy** [comment]
@awni ready for review.

File: llms/mlx_lm/models/qwen2_moe.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+@dataclass
+class ModelArgs(BaseModelArgs):
+    model_type: str
+    hidden_size: int
+    num_hidden_layers: int
+    intermediate_size: int
+    num_attention_heads: int
+    num_experts_per_tok: int
+    num_experts: int
+    decoder_sparse_step: int
```
Comment (awni) [inline llms/mlx_lm/models/qwen2_moe.py:20]: This argument is not used anywhere as far as I can tell. Can you remove?

Comment (Blaizzy) [reply llms/mlx_lm/models/qwen2_moe.py]: Removed it ✅ 

It was being used in my version but now it's not.

File: llms/mlx_lm/utils.py [hunk 4]
```diff
@@ -35,6 +35,8 @@
     lambda m: isinstance(m, nn.Linear)
     and m.weight.shape[0]
     != 8  # avoid quantizing gate layers, otherwise we have to re-quant and upload all the mixtral models
+    and m.weight.shape[0]
+    not in [1, 60] # avoid quantizing qwen2_moe gate layers
```
[the remark this answers is not in the document]
Comment (Blaizzy) [reply llms/mlx_lm/utils.py:39]: It's working fine without it but gimme a few minutes to test it.

[the remark this answers is not in the document]
Comment (Blaizzy) [reply llms/mlx_lm/utils.py:39]: There is no difference in the response or speed so I think we can remove it.

[the remark this answers is not in the document]
Comment (awni) [reply llms/mlx_lm/utils.py:39]: Great! Let me know when it's done then we can merge it!

[the remark this answers is not in the document]
Comment (Blaizzy) [reply llms/mlx_lm/utils.py]: Remove it ✅

**Comment by awni** [comment]
Could you check the test @Blaizzy ?

**Comment by Blaizzy** [comment]
> Could you check the test @Blaizzy ? Looks like you forgot to remove a field from there.

Sure let me check

**Comment by Blaizzy** [comment]
Fixed it ✅

**Review by awni — APPROVED** [review]

**Comment by awni** [comment]
Thanks @Blaizzy and @bozheng-hit for the addition!

**Comment by Blaizzy** [comment]
Most welcome, it's my pleasure :)

Thank you @bozheng-hit for the addition!

File: llms/mlx_lm/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bozheng-hit) [reply llms/mlx_lm/utils.py]: I suggest not to quantizing the gating layers since the routing function is sensitive, unless you can prove that it won't affect downstream tasks at all. I did not quantize these layers in the GPTQ-int4 version.

[the remark this answers is not in the document]
Comment (awni) [reply llms/mlx_lm/utils.py]: Would be great to see some results on that.. of course you always lose a little from quantizing any matrix but are the results more sensitive to quantizing the gate than other matrices?  That would be very useful to know if it's generally the case. 

(I believe @mzbac has possibly observed this as well)
